site/docs/tutorial/sql_to_substrait.md

```diff
@@ -0,0 +1,1305 @@
+                  "scalarFunction": {
+                    "functionReference": 3,
+                    "outputType": {
+                      "decimal(38, 2)": {
+                        "nullability": "NULLABILITY_NULLABLE"
+                      },
+                      "arguments": [
+                        {
+                          "enum": {
+                            "specified": "OVERFLOW_ERROR"
+                          }
+                        },
+                        {
+                          "value": {
+                            "cast": {
+                              "type": {
+                                "decimal(10, 2)": {
+                                  "nullability": "NULLABILITY_REQUIRED"
+                                }
+                              },
+                              "input": {
+                                "selection": {
+                                  "directReference": {
+                                    "structField": {
+                                      "field": 1
+                                    }
+                                  }
+                                }
+                              }
+                            }
+                          }
+                        },
+                        {
+                          "value": {
+                            "selection": {
+                              "directReference": {
+                                "structField": {
+                                  "field": 3
+                                }
+                              }
+                            }
+                          }
+                        }
+                      ]
+                    }
+                  }
+                }
+              }
+            ]
+          }
+        }
+      }
+    ]
+  }
+}
+```
+
+## Field indices
+
+So far, we have glossed over the field indices. Now that we've built up each
+of the relations, it will be a bit easier to explain them.
+
+Instead of referencing fields by name, in Substrait we always reference them by
+index. Nested fields are part of the index sequence, put in depth-first order.
+For example, as shown in the diagram below, in the `products` table read relation
+the `manufacturer` and `year_created` fields come right after their parent field,
+`details`.
+
+One complication is that whenever we reference by field index, those field
+indices are dependent on the context of the relation. The exact field indices
+can change from relation to relation. For example, a join relations field 
+indices are based on the combination of the left inputs fields and then the 
+right input fields. And, while we don't have one in our example, a Project relation
+adds new fields, which are visible to it's downstream relations. The diagram
+below shoes the mapping of field indices within each relation and how each of
+the field references show up in each relations properties.
```

Comment:
Have I described the rules adequately / accurately? Is there anything important I've missed?

Reply:
> those field indices are dependent on the context of the relation

> a join relations field indices

If I were to be pedantic I would say "the context of the expression".  A join relation does not have field indices.  A join relation has a "join expression" and the context of references in that expression is the output schema of the join relation.

An interesting case to consider (and I have a really old PR I need to get merged clarifying this :grimacing: https://github.com/substrait-io/substrait/pull/271/files) is the read relation.

A read relation has a filter expression and a projection (which isn't an expression exactly, but a masked expression).  A read relation effectively has _two_ schemas (maybe even more if we want to consider evolution but we won't go there :laughing: ).  There is the schema of the data on disk (the direct schema) and the output schema of the read relation (the direct schema after the projection mask has been applied to it).  So we have to be specific when talking about the filter expression which schema we are actually binding against.

Reply:
Thanks. It is complex, but I found that `substrait-validator` provides a really helpful UI for showing those schemas and so I've referenced that tool in the tutorial.